### PR TITLE
feat: add task update handler and util

### PR DESCRIPTION
**Pull request by tejaskh3**
Date: 2 April. 2024

Developer Name: Tejas

---

## Description
Developing an API that sends the updates of tasks in the RDS discord channel whenever the user updates his/her progress.

changes made in PR are firstly creating a controller that authorizes if the request is only coming from the backend of RDS  and then passing it to the util function which sends an update message on the tracking channel if the bot is authorized then only.

### Documentation Updated?

- [ ] Yes
- [x] No

### Under Feature Flag

- [ ] Yes
- [x] No

### Database Changes

- [ ] Yes
- [x] No

### Breaking Changes

- [ ] Yes
- [x] No

### Development Tested?

- [x] Yes
- [ ] No

## Screenshots
<details>
<summary>Screenshot 1</summary>

https://github.com/Real-Dev-Squad/discord-slash-commands/assets/98630752/7ccfa065-dcea-4614-bc61-8c493b77bda1

</details>

## Test Coverage
image

image

image

</details>

### CODE REVIEW

File: tests/unit/utils/sendTasksUpdates.test.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import JSONResponse from "../../../src/utils/JsonResponse";
+import { sendTaskUpdate } from "../../../src/utils/sendTaskUpdates";
+
+describe("sendTaskUpdate function", () => {
+  const mockEnv = { DISCORD_TOKEN: "mockToken" };
+  const completed = "Task completed successfully";
+  const planned = "Plan for the next phase";
+  const blockers = "No blockers";
+
+  test("should send the task update to discord tracking channel.", async () => {
```
Comment (lakshayman) [inline tests/unit/utils/sendTasksUpdates.test.ts:11]: You can write tests for the undefined body keys. Like undefined content, undefined blockers, undefined completed, etc.

Comment (tejaskh3) [reply tests/unit/utils/sendTasksUpdates.test.ts:11]: Okay, I'll write that.

Comment (tejaskh3) [reply tests/unit/utils/sendTasksUpdates.test.ts:11]: Done sir, please check.

File: src/controllers/taskUpdatesHandler.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    }
+    await verifyNodejsBackendAuthToken(authHeader, env);
+    const updates: {
+      content: { completed: string; planned: string; blockers: string };
+    } = await request.json();
+    const { completed, planned, blockers } = updates.content;
+    await sendTaskUpdate(completed, planned, blockers, env);
+    return new JSONResponse(
+      "Task update sent on discord tracking updates channel."
+    );
```
Comment (ankushdharkar) [inline src/controllers/taskUpdatesHandler.ts:21]: Is the try-catch correctly performed here? Seems like any exception will end up returning bad signature reason, and that may be misleading.

**Review by ankushdharkar — CHANGES_REQUESTED** [review]

File: src/controllers/taskUpdatesHandler.ts (hunk 2 above)
Comment (tejaskh3) [reply src/controllers/taskUpdatesHandler.ts:21]: Yes, this should be fixed to an internal server error.
Please check now.

File: src/utils/sendTaskUpdates.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  await fetch(url, {
+    method: "POST",
+    body: JSON.stringify(bodyObj),
+    headers: {
+      "Content-Type": "application/json",
+      Authorization: `Bot ${env.DISCORD_TOKEN}`,
+    },
+  });
+  return;
+}
```
Comment (Achintya-Chatterjee) [inline src/utils/sendTaskUpdates.ts:24]: `sendTaskUpdate` function does not handle errors from the fetch call. It's important to handle potential network errors, such as timeouts or failed requests, to provide better resilience to the application.

File: src/controllers/taskUpdatesHandler.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+export const sendTaskUpdatesHandler = async (request: any, env: env) => {
+  try {
+    const authHeader = request.headers.get("Authorization");
+    if (!authHeader) {
+      return new JSONResponse(response.BAD_SIGNATURE);
+    }
+    await verifyNodejsBackendAuthToken(authHeader, env);
+    const updates: {
+      content: { completed: string; planned: string; blockers: string };
```
Comment (Achintya-Chatterjee) [inline src/controllers/taskUpdatesHandler.ts:15]: When returning an error response in the `catch` block of `sendTaskUpdatesHandler`, the `message` field includes the error object directly. Extracting relevant information from the error object or providing a more user-friendly error message is recommended.

File: tests/unit/utils/sendTasksUpdates.test.ts
Comment (Achintya-Chatterjee) [inline tests/unit/utils/sendTasksUpdates.test.ts:1]: unit tests are there for the `sendTaskUpdate` function, it's important to ensure comprehensive test coverage for all code paths, including error scenarios and edge cases.

File: src/utils/sendTaskUpdates.ts (hunk 3 above)
Comment (tejaskh3) [reply src/utils/sendTaskUpdates.ts:24]: Please check the error handling is done in the taskUpdateHandler.

File: src/controllers/taskUpdatesHandler.ts (hunk 4 above)
Comment (tejaskh3) [reply src/controllers/taskUpdatesHandler.ts:15]: I have already updated that on @ankushdharkar 's comment.

File: src/controllers/taskUpdatesHandler.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    } = await request.json();
```
Comment (yesyash) [inline src/controllers/taskUpdatesHandler.ts:16]: - create a type for this above the `sendTaskUpdatesHandler` as it will allow us to reuse this type in different methods if required.

File: src/controllers/taskUpdatesHandler.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
+import { env } from "../typeDefinitions/default.types";
+import JSONResponse from "../utils/JsonResponse";
+import * as response from "../constants/responses";
+import { verifyNodejsBackendAuthToken } from "../utils/verifyAuthToken";
+import { sendTaskUpdate } from "../utils/sendTaskUpdates";
+
+export const sendTaskUpdatesHandler = async (request: any, env: env) => {
```
Comment (yesyash) [inline src/controllers/taskUpdatesHandler.ts:7]: - please change the request type from `any` to a more narrowed type.

**Review by yesyash — CHANGES_REQUESTED** [review]

File: src/controllers/taskUpdatesHandler.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+  try {
+    const authHeader = request.headers.get("Authorization");
+    if (!authHeader) {
+      return new JSONResponse(response.BAD_SIGNATURE);
+    }
```
[the remark this answers is not in the document]
Comment (tejaskh3) [reply src/controllers/taskUpdatesHandler.ts]: No, that is not covered in the tests. I am trying to do that.

the status code and message have been updated and tested please check.
Screenshot 2024-04-08 164553

File: src/controllers/taskUpdatesHandler.ts (hunk 5 above)
Comment (tejaskh3) [reply src/controllers/taskUpdatesHandler.ts]: Done.

File: src/controllers/taskUpdatesHandler.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (tejaskh3) [reply src/controllers/taskUpdatesHandler.ts]: Now it is covered in the test cases please check.

File: src/controllers/taskUpdatesHandler.ts (hunk 6 above)
Comment (tejaskh3) [reply src/controllers/taskUpdatesHandler.ts]: Done

File: src/controllers/taskUpdatesHandler.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+  if (!authHeader) {
+    return new JSONResponse(response.UNAUTHORIZED, { status: 401 });
+  }
+  try {
+    await verifyNodejsBackendAuthToken(authHeader, env);
+    const updates: TaskUpdates = await request.json();
+    const { completed, planned, blockers } = updates.content;
+    await sendTaskUpdate(completed, planned, blockers, env);
+    return new JSONResponse(
+      "Task update sent on discord tracking updates channel."
```
Comment (iamitprakash) [inline src/controllers/taskUpdatesHandler.ts:20]: Task update sent on Discord's tracking-updates channel.

**Review by iamitprakash — CHANGES_REQUESTED** [review]

File: src/utils/sendTaskUpdates.ts [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+  const url = config(env).TRACKING_CHANNEL_URL;
+  await fetch(url, {
+    method: "POST",
+    body: JSON.stringify(bodyObj),
+    headers: {
+      "Content-Type": "application/json",
+      Authorization: `Bot ${env.DISCORD_TOKEN}`,
+    },
+  });
+  return;
```
Comment (iamitprakash) [inline src/utils/sendTaskUpdates.ts:23]: Explicit Return:  Add return Promise.resolve(); at the end to indicate successful execution.

Error Handling: Consider adding error handling (e.g., using a try...catch block) to gracefully handle potential issues during the fetch operation.

**Review by iamitprakash — CHANGES_REQUESTED** [review]

**Review by iamitprakash — APPROVED** [review]

**Review by ardourApeX — APPROVED** [review]
